### PR TITLE
Repopulate cached GlobalConfig object if CYLC_SYMLINKS exported to env.

### LINKED ISSUE #237: Environment variables set in rose-suite.conf should apply when the global config is loaded
### Problem

Some users want to be able to choose a different symlink dirs setup per workflow, see https://github.com/cylc/cylc-flow/issues/5418.
This can be done via the use of environment variables and Jinja2 in the global config, e.g.:
```
#!jinja2
[install]
    [[symlink dirs]]
        [[[hpc]]]
{% if environ["CYLC_SYMLINKS"] | default("x") == "A" %}
            run = $PROJECT_A
{% elif environ["CYLC_SYMLINKS"] | default("x") == "B" %}
            run = $PROJECT_B
{% else %}
            run = $DATADIR
{% endif %}
```

Where appropriate (depending on the use case) this could be configured at the site level or the user level.

In order for this to work you need to be able to configure an environment variable to be set both for `cylc install` (for localhost) and for `cylc play` (for remote hosts). It is already possible to define environment variables in `rose-suite.conf` which alter the scheduler environment:
https://metomi.github.io/rose/doc/html/api/configuration/suite.html#rose:conf:rose-suite.conf[env]
However, at present these do not apply when the global config is loaded.

### Proposed Solution

Ensure that environment variables set in `rose-suite.conf` apply when the global config is loaded (or reloaded if necessary).

### PR DESCRIPTION
**Pull request by wxtim**
Closes #237 

This is now ready for review but needs testing against https://github.com/cylc/cylc-flow/pull/5839.

Ensure that the cached globalconfig object is reloaded after the export of `CYLC_SYMLINKS` variable. This means that using the `CYLC_SYMLINKS` variable allows users to specify installation symlink locations in the `rose-suite.conf`.

n.b. I was surprised to find that running `glbl_cfg()._DEFAULT = None` did not clear the cached config in a way that would force the reload in Cylc, but I think just reloading the config in Cylc Rose makes at least as much sense.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] Docs: I have updated the plugin documentation.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -6,6 +6,14 @@ creating a new release entry be sure to copy & paste the span tag with the
 updated. Only the first match gets replaced, so it's fine to leave the old
 ones in. -->
 
+
+## __cylc-rose-1.4.0 (<span actions:bind='release-date'>Upcoming</span>)__
+
+### Features
+
+[#269](://github.com/cylc/cylc-rose/pull/269) - Allow Environment variables
```
Comment (MetRonnie) [inline CHANGES.md:14]:
```suggestion
[#269](https://github.com/cylc/cylc-rose/pull/269) - Allow environment variables
```

File: cylc/rose/__init__.py
Comment (MetRonnie) [inline cylc/rose/__init__.py:1]: I think Oliver has said there is no special handling for `CYLC_SYMLINKS`?

**Comment by wxtim** [comment]
Now failing identically to [master nightly tests](https://github.com/cylc/cylc-rose/actions/runs/7381378751/job/20079851194).

**Review by oliver-sanders — APPROVED** [review]
LGTM

File: tests/functional/test_utils.py [hunk 2]
```diff
@@ -36,3 +36,84 @@ def test_basic(tmp_path):
...
+        '#!jinja2\n'
+        '[install]\n'
+        '    [[symlink dirs]]\n'
+        '        [[[localhost]]]\n'
+        '{% set cylc_symlinks = environ.get(\'CYLC_SYMLINKS\', None) %}\n'
+        '{% if cylc_symlinks == "foo" %}\n'
+        f'log = {str(tmp_path)}/foo\n'
+        '{% else %}\n'
+        f'log = {str(tmp_path)}/bar\n'
+        '{% endif %}\n'
```
[the remark this answers is not in the document]
Comment (wxtim) [reply tests/functional/test_utils.py:94]: I agree, but escaping the Jinja2 would be a pain, where here I just make the lines with Jinja2 non-f-strings.

File: cylc/rose/__init__.py [hunk 3]
```diff
@@ -70,11 +70,37 @@
...
+      [install]
+         [[symlink dirs]]
+            [[[hpc]]]
+      {% if environ["CYLC_SYMLINKS"] | default("x") == "A" %}
+               run = $LOCATION_A
+      {% elif environ["CYLC_SYMLINKS"] | default("x") == "B" %}
+               run = $LOCATION_B
+      {% else %}
+               run = $LOCATION_C
+      {% endif %}
```
Comment (MetRonnie) [inline cylc/rose/__init__.py:95]:
```suggestion
.. code-block:: cylc

   #!jinja2
   # part of a global.cylc file
   [install]
      [[symlink dirs]]
         [[[hpc]]]
   {% if environ["CYLC_SYMLINKS"] | default("x") == "A" %}
            run = $LOCATION_A
   {% elif environ["CYLC_SYMLINKS"] | default("x") == "B" %}
            run = $LOCATION_B
   {% else %}
            run = $LOCATION_C
   {% endif %}
```

File: cylc/rose/__init__.py [hunk 4]
```diff
@@ -70,11 +70,37 @@
...
+
+
+
 Special Variables
 -----------------
 
 The Cylc Rose plugin provides two environment/template variables
```
Comment (MetRonnie) [inline cylc/rose/__init__.py:102]: (unrelated to your changes, but just spotted)
```suggestion
The Cylc Rose plugin provides three environment/template variables
```

File: tests/conftest.py [hunk 5]
```diff
@@ -119,16 +131,21 @@ def _inner(srcpath, args=None):
         """
         options = Options(install_gop(), args)()
         output = SimpleNamespace()
+        if not options.workflow_name:
+            options.workflow_name = workflow_name
+        if not args or args and not args.get('no_run_name', ''):
```
Comment (MetRonnie) [inline tests/conftest.py:136]:
```suggestion
        if not args or not args.get('no_run_name', ''):
```

File: tests/functional/test_utils.py [hunk 6]
```diff
@@ -36,3 +36,92 @@ def test_basic(tmp_path):
...
+            [[x]]
+    """)
+
+    # Validate the config:
+    output = cylc_validate_cli(tmp_path)
+    assert output.ret == 0
+
+    # CYLC_SYMLINKS == None the first time the global.cylc
+    # is loaded and "Foo" the second time.
+    assert output.logging == '"Foo"'
```
Comment (MetRonnie) [inline tests/functional/test_utils.py:75]: Running locally I get
```
        # CYLC_SYMLINKS == None the first time the global.cylc
        # is loaded and "Foo" the second time.
>       assert output.logging == '"Foo"'
E       assert 'None\n"Foo"' == '"Foo"'
E         + None
E           "Foo"

tests/functional/test_utils.py:75: AssertionError
```

**Comment by wxtim** [comment]
@MetRonnie  - this is failing in the same way as master.

**Review by MetRonnie — APPROVED** [review]

File: cylc/rose/utilities.py [hunk 7]
```diff
@@ -869,6 +870,13 @@ def export_environment(environment: Dict[str, str]) -> None:
     for key, val in environment.items():
         os.environ[key] = val
 
+    # If env vars have been set we want to force reload
+    # the global config so that the value of this vars
+    # can be used by Jinja2 in the global config.
+    # https://github.com/cylc/cylc-rose/issues/237
+    if environment:
```
Comment (oliver-sanders) [inline cylc/rose/utilities.py:877]: Note, the environment is always populated as a result of automatic Rose variables.

**Pull request merged**